conjure-core/src/main/java/com/palantir/conjure/defs/EndpointErrorResolver.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.defs;
+
+import com.palantir.conjure.parser.types.ConjureType;
+import com.palantir.conjure.parser.types.ConjureTypeVisitor;
+import com.palantir.conjure.parser.types.builtin.AnyType;
+import com.palantir.conjure.parser.types.builtin.BinaryType;
+import com.palantir.conjure.parser.types.builtin.DateTimeType;
+import com.palantir.conjure.parser.types.collect.ListType;
+import com.palantir.conjure.parser.types.collect.MapType;
+import com.palantir.conjure.parser.types.collect.OptionalType;
+import com.palantir.conjure.parser.types.collect.SetType;
+import com.palantir.conjure.parser.types.complex.ErrorTypeDefinition;
+import com.palantir.conjure.parser.types.names.ConjurePackage;
+import com.palantir.conjure.parser.types.names.TypeName;
+import com.palantir.conjure.parser.types.primitive.PrimitiveType;
+import com.palantir.conjure.parser.types.reference.ForeignReferenceType;
+import com.palantir.conjure.parser.types.reference.LocalReferenceType;
+import com.palantir.logsafe.Preconditions;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Endpoint error definitions are either references to errors defined in the same file as the endpoint definition, or in
+ * imported Conjure files. This class resolved the reference to the error type populating the package name if necessary.
+ */
+public final class EndpointErrorResolver {
+    private final Map<TypeName, ErrorTypeDefinition> parsedErrors;
+    private final Optional<ConjurePackage> defaultConjurePackage;
+
+    public EndpointErrorResolver(
+            Map<TypeName, ErrorTypeDefinition> parsedErrors, Optional<ConjurePackage> defaultConjurePackage) {
+        this.parsedErrors = parsedErrors;
+        this.defaultConjurePackage = defaultConjurePackage;
+    }
+
+    com.palantir.conjure.spec.TypeName resolve(ConjureType conjureType) {
+        Optional<TypeName> maybeTypeName = conjureType.visit(GetReferenceTypeNameVisitor.INSTANCE);
+        Preconditions.checkArgument(
+                maybeTypeName.isPresent(),
+                "Unsupported endpoint error type. Endpoint errors must be references to a Conjure-defined"
+                        + " error type");
```

Comment:
if we throw the exception from `GetReferenceTypeNameVisitor` , we can provide more precise information. Thoughts on updating from `Optional<TypeName>` to `TypeName` and throwing in the cases that currently return `Optional.empty()`?

Reply:
Yeah that makes sense. The version you reviewed actually did not handle imported errors correctly. I've updated the logic to handle them correctly, and to expose the `type` in the IllegalArgumentException thrown when it's not one of the reference types.